File: accidentes/FicheroAccidentes-Java-IvanRC/src/main/java/storage/accidentes/AccidenteStorageServiceJson.java
package storage.accidentes;

import com.squareup.moshi.Moshi;
import config.ConfigApp;
import dto.conjuntoDeAccidentes.ConjuntosDeAccidentesDto;
import mapper.accidente.AccidenteMapper;
import model.Accidente;

import java.io.*;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

public class AccidenteStorageServiceJson implements AccidenteStorageService{

    ConfigApp configApp = ConfigApp.getInstance();

    File file = new File(configApp.getAPP_DATA()+File.separator+"accidentes.json");

    Moshi moshi = new Moshi.Builder().build();

    public AccidenteStorageServiceJson() throws IOException {}

    @Override
    public void saveAll(List<Accidente> entities) throws Exception {
        try(BufferedWriter br = new BufferedWriter(new FileWriter(file))){
            br.write(moshi.adapter(ConjuntosDeAccidentesDto.class).indent("   ").toJson(new ConjuntosDeAccidentesDto(AccidenteMapper.toAccidentesDto(entities))));
        }catch (Exception e){
            throw e;
        }
    }

    @Override
    public List<Accidente> loadAll() throws Exception {
        List<Accidente> lista = new ArrayList<>();
        try(BufferedReader br = new BufferedReader(new FileReader(file))) {
            lista = AccidenteMapper.toAccidentes(moshi.adapter(ConjuntosDeAccidentesDto.class).fromJson(Files.readString(Path.of(file.getPath()))).getAccidentesDto());
        }catch (Exception e){
            throw e;
        }
        return lista;
    }
}

File: accidentes/FicheroAccidentes-Kotlin-IvanRC/src/main/kotlin/storage/accidentes/AccidenteStorageServiceJSON.kt
package storage.accidentes

import com.squareup.moshi.Moshi
import com.squareup.moshi.adapter
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import config.ConfigApp
import dto.conjuntoDeAccidentes.ConjuntoDeAccidenteDto
import mapper.accidente.toAccidente
import mapper.accidente.toAccidentes
import mapper.accidente.toAccidentesDto
import mapper.accidente.toDto
import model.Accidente
import mu.KotlinLogging
import java.io.File

@ExperimentalStdlibApi
class AccidenteStorageServiceJSON : AccidenteStorageService{

    private val configApp = ConfigApp
    private val file = File(configApp.APP_DATA+File.separator+"accidentes.json")
    private val logger = KotlinLogging.logger {}

    private val moshi = Moshi.Builder().addLast(KotlinJsonAdapterFactory()).build()
    val jsonAdapter = moshi.adapter<ConjuntoDeAccidenteDto>()

    override fun saveAll(entities: List<Accidente>) {
        logger.debug { "Se guardan todos los accidentes en el fichero JSON." }
        file.writeText(jsonAdapter.indent("   ").toJson(ConjuntoDeAccidenteDto(entities.toAccidentesDto())))
    }

    override fun loadAll(): List<Accidente> {
        logger.debug { "Se cargan todos los accidentes del fichero JSON." }
        if(!file.exists()) return emptyList()
        return jsonAdapter.fromJson(file.readText())!!.accidentes.toAccidentes()
    }

}